### PR TITLE
User middleware

### PR DESCRIPTION
**Pull request by bergie**
Moves user data handling from mixture of storage and UI to a Redux-style middleware graph:

screenshot 2017-01-23 at 17 39 17

Middleware components receive an action, and can either pass the action onwards, or capture it and emit their own actions instead.

This way for instance a `user:login` action gets caught by the UserMiddleware graph, which then performs the OAuth dance, and down the line emits either `user:info` (for successful login) or `user:error` (for failed login).

### CODE REVIEW

File: components/SetAction.coffee [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  c.outPorts.add 'out',
+    datatype: 'all'
+
+  noflo.helpers.WirePattern c,
+    params: ['action']
+    forwardGroups: false
+    async: true
+  , (data, groups, out, callback) ->
+    actionParts = c.params.action.split ':'
+    setTimeout ->
```
Comment (jonnor) [inline components/SetAction.coffee]: Why `setTimeout`?

File: components/SetAction.coffee [hunk 2]
```diff
@@ -0,0 +1,25 @@
+noflo = require 'noflo'
+
+exports.getComponent = ->
+  c = new noflo.Component
+  c.inPorts.add 'action',
+    datatype: 'string'
+    required: true
+    control: true
```
Comment (jonnor) [inline components/SetAction.coffee]: Why `control: true` ?

File: components/SetUrl.coffee [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  c.inPorts.add 'in',
+    datatype: 'string'
+  c.outPorts.add 'out',
+    datatype: 'bang'
+
+  noflo.helpers.WirePattern c,
+    async: true
+    forwardGroups: false
+  , (data, groups, out, callback) ->
+    # This will in effect cause a NoFlo network stop as the app
```
Comment (jonnor) [inline components/SetUrl.coffee]: Can't we use some single-page-app magic to avoid this? For future..

File: spec/browser.html [hunk 4]
```diff
@@ -28,11 +28,11 @@
   <body>
     <div id="mocha"></div>
     <iframe id="app"></iframe>
-    <script src="Initialization.js"></script>
-    <script src="Login.js"></script>
-    <script src="CreateProject.js"></script>
-    <script src="EditGraph.js"></script>
-    <script src="DeleteGraph.js"></script>
+    <script src="EndToEndInitialization.js"></script>
```
Comment (jonnor) [inline spec/browser.html]: Can we use a directory for these tests? `spec/EndToEndLogin.coffee`.

File: package.json [hunk 5]
```diff
@@ -17,14 +17,15 @@
...
+    "debug": "^2.6.0",
     "fbp": "~1.5.0",
     "fbp-spec": "~0.1.15",
     "flowhub-registry": "^0.1.0",
     "font-awesome": "^4.6.3",
     "indexeddbshim": "^2.2.1",
     "jshint": "^2.9.2",
     "microflo": "~0.3.43",
-    "noflo": "~0.7.9",
+    "noflo": "^0.7.9",
```
Comment (jonnor) [inline package.json]: On the 0.x series these are effectively the same in semver. Why the change?

File: components/LoadUserData.coffee [hunk 6]
```diff
@@ -46,5 +46,3 @@ exports.getComponent = ->
       return callback err if err
       out.send valid
       do callback
-
-  c
```
Comment (jonnor) [inline components/LoadUserData.coffee]: Bug? Or does WirePattern return the component?

File: spec/UrlMiddleware.coffee [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    check = (data) ->
+      # Strict equality check for passed packets
+      chai.expect(data).to.equal payload
+    receiveAction socket, action, check, done
+
+  describe 'receiving a runtime:connect action', ->
+    it 'should pass it out as-is', (done) ->
+      action = 'runtime:connect'
+      payload =
+        hello: 'world'
```
Comment (jonnor) [inline spec/UrlMiddleware.coffee]: Eh, what is this case for? Does not seem useful

File: spec/UserMiddleware.coffee [hunk 8]
```diff
@@ -0,0 +1,258 @@
+noflo = require 'noflo'
+
+unless noflo.isBrowser()
+  chai = require 'chai'
+  path = require 'path'
+  baseDir = path.resolve __dirname, '../'
+else
+  baseDir = 'noflo-ui'
+
+describe 'User Middleware', ->
```
Comment (jonnor) [inline spec/UserMiddleware.coffee]: Yay tests!

File: spec/UserMiddleware.coffee [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+    loader = new noflo.ComponentLoader baseDir
+    loader.load 'ui/UserMiddleware', (err, instance) ->
+      return done err if err
+      c = instance
+      actionIn = noflo.internalSocket.createSocket()
+      c.inPorts.action.attach actionIn
+      actionIn.port = 'action'
+      ###
+      c.network.on 'begingroup', (data) ->
+        console.log "   < #{data.id}"
```
Comment (jonnor) [inline spec/UserMiddleware.coffee]: Begging for flowtrace...

File: spec/UserMiddleware.coffee [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+    describe 'without user and with grant code in URL yielding invalid API token', ->
+      xhr = null
+      code = null
+      token = null
+      beforeEach ->
+        code = 'oivwehfh24890f84h'
+        token = 'niov2o3wnnv4ioufuhh92348fh42q9'
+        xhr = sinon.fakeServer.create()
+      afterEach ->
+        xhr.restore()
```
Comment (jonnor) [inline spec/UserMiddleware.coffee]: Would prefer to use a name for the mock which singifies testing more than `xhr`. Like `mock` or `mockAuth`

File: spec/UserMiddleware.coffee [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+          , JSON.stringify
+            token: token
+        xhr.respondWith 'GET', "https://api.flowhub.io/user", (req) ->
+          req.respond 401,
+            'Content-Type': 'application/json'
+          , JSON.stringify
+            message: 'Bad Credentials'
+        do xhr.respond
+        do xhr.respond
+    describe 'without user and with valid grant code in URL', ->
```
Comment (jonnor) [inline spec/UserMiddleware.coffee]: Should this be `with user` instead of `without`?

**Comment by jonnor** [comment]
No unused components to delete after this?

**Comment by jonnor** [comment]
I think the `action` outputs on the middlewares should be something like `new` (or `created` or similar), to better clarify their semantics and match `pass`.

**Comment by jonnor** [comment]
@bergie I don't understand how this is different from the existing "Store" code... I don't see any removed code from UI layer?

File: components/SetUrl.coffee (hunk 3 above)
Comment (bergie) [inline components/SetUrl.coffee]: We're using this to redirect to GitHub, too. The URL query cleanup we do post-login can be done down the line with pushState

File: package.json (hunk 5 above)
Comment (bergie) [inline package.json]: I think side effect of trying 0.8 beta and then going back to 0.7 to get only one NoFlo into the bundle (as opposed to N pulled in via individual modules)

File: components/LoadUserData.coffee (hunk 6 above)
Comment (bergie) [inline components/LoadUserData.coffee]: WirePattern returns the component, just like most of the component methods

File: spec/UserMiddleware.coffee (hunk 9 above)
Comment (bergie) [inline spec/UserMiddleware.coffee]: No flowtraces on bare (non-runtime) NoFlo networks I think? Would be super useful

File: components/SetAction.coffee (hunk 2 above)
Comment (bergie) [inline components/SetAction.coffee]: Not really needed, was going to do it with process API first, but decided to stick to WirePattern until 0.8 stable is out

File: components/SetAction.coffee (hunk 1 above)
Comment (bergie) [inline components/SetAction.coffee]: Workaround for bracket forwarding bug in WirePattern. Would be fixed by going 0.8 process API

File: spec/UrlMiddleware.coffee (hunk 7 above)
Comment (bergie) [inline spec/UrlMiddleware.coffee]: I wanted a test of a passed packet. The contents don't really matter as long as:

* the action is one that the middleware isn't supposed to handle
* we verify the packet gets passed through without manipulation (strict equality check)

**Comment by bergie** [comment]
`RemoteLogin` and `UserStorage` components were removed in this PR (RemoteLogin was split to multiple components for better testability).

No code removal from UI yet, though now that we have all of this in place removing the user API calls in `noflo-account-settings` is a fast follow

**Comment by jonnor** [comment]
@bergie ok looks good to me. Merge at will
